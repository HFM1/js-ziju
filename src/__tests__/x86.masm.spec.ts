import {
  Assembler
} from '../index'
import { build, readSourceWithKernal } from './utils'

describe('compiler x86 masm', () => {
  it('print', () => {
    let result = new Assembler(readSourceWithKernal('tests/asm/print.js')).compile()

    expect(result.assembly).toMatchSnapshot()

    build(result.assembly,'print')
  });

  it('functionCall', () => {
    let result = new Assembler(readSourceWithKernal('tests/asm/functionCall.js')).compile()

    expect(result.assembly).toMatchSnapshot()

    build(result.assembly,'functionCall')
  });

  it('fib', () => {
    let result = new Assembler(readSourceWithKernal('tests/asm/fib.js')).compile()

    expect(result.assembly).toMatchSnapshot()

    build(result.assembly,'fib')
  });
  
  it('tail-fib', () => {
    let result = new Assembler(readSourceWithKernal('tests/asm/tail-fib.js')).compile()

    expect(result.assembly).toMatchSnapshot()

    build(result.assembly,'tail-fib')
  });
});
