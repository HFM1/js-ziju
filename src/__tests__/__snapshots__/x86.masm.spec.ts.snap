// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler x86 masm fib 1`] = `
"  .global _main

  .text

printChar:
  PUSH RBP
  MOV RBP, RSP

  PUSH 1
  MOV RAX, RBP
  ADD RAX, 16 # $$c
  PUSH RAX
  PUSH 1
  POP RDX
  POP RSI
  POP RDI
  MOV RAX, 33554436
  SYSCALL
  PUSH RAX

  POP RAX
  POP RBP

  RET

printHelper:
  PUSH RBP
  MOV RBP, RSP

  # If
    # >
    PUSH [RBP + 16] # n
    PUSH 9
    POP RAX
    CMP [RSP], RAX
    MOV RAX, 0
    MOV DWORD PTR [RSP], 1
    CMOVA RAX, [RSP]
    MOV [RSP], RAX
    # End >

  POP RAX
  TEST RAX, RAX
  JZ .else_branch3

  # If then
    # DIV
    PUSH [RBP + 16] # n
    PUSH 10
    POP RAX
    XCHG [RSP], RAX
    XOR RDX, RDX
    DIV QWORD PTR [RSP]
    MOV [RSP], RAX
  CALL printHelper
  MOV [RSP], RAX

  JMP .after_else_branch3

  # If else
.else_branch3:
  PUSH 0 # Null else branch
.after_else_branch3:
  # End if
  POP RAX # Ignore non-final expression
    # ADD
    PUSH 48
      # DIV
      PUSH [RBP + 16] # n
      PUSH 10
      POP RAX
      XCHG [RSP], RAX
      XOR RDX, RDX
      DIV QWORD PTR [RSP]
      MOV [RSP], RDX
    POP RAX
    ADD [RSP], RAX
    # End ADD
  CALL printChar
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

print:
  PUSH RBP
  MOV RBP, RSP

  PUSH [RBP + 16] # n
  CALL printHelper
  MOV [RSP], RAX

  POP RAX # Ignore non-final expression
  PUSH 10
  CALL printChar
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

fib:
  PUSH RBP
  MOV RBP, RSP

  # If
    # <
    PUSH [RBP + 16] # n
    PUSH 2
    POP RAX
    CMP [RSP], RAX
    MOV RAX, 0
    MOV DWORD PTR [RSP], 1
    CMOVB RAX, [RSP]
    MOV [RSP], RAX
    # End <

  POP RAX
  TEST RAX, RAX
  JZ .else_branch4

  # If then
  PUSH [RBP + 16] # n
  JMP .after_else_branch4

  # If else
.else_branch4:
    # ADD
      # SUB
      PUSH [RBP + 16] # n
      PUSH 1
      POP RAX
      SUB [RSP], RAX
      # End SUB
    CALL fib
    MOV [RSP], RAX

      # SUB
      PUSH [RBP + 16] # n
      PUSH 2
      POP RAX
      SUB [RSP], RAX
      # End SUB
    CALL fib
    MOV [RSP], RAX

    POP RAX
    ADD [RSP], RAX
    # End ADD
.after_else_branch4:
  # End if
  POP RAX
  POP RBP

  RET

main:
  PUSH RBP
  MOV RBP, RSP

  PUSH 5
  CALL fib
  MOV [RSP], RAX

  CALL print
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

_main:
  CALL main
  MOV RDI, RAX
  MOV RAX, 33554433
  SYSCALL
"
`;

exports[`compiler x86 masm functionCall 1`] = `
"  .global _main

  .text

printChar:
  PUSH RBP
  MOV RBP, RSP

  PUSH 1
  MOV RAX, RBP
  ADD RAX, 16 # $$c
  PUSH RAX
  PUSH 1
  POP RDX
  POP RSI
  POP RDI
  MOV RAX, 33554436
  SYSCALL
  PUSH RAX

  POP RAX
  POP RBP

  RET

printHelper:
  PUSH RBP
  MOV RBP, RSP

  # If
    # >
    PUSH [RBP + 16] # n
    PUSH 9
    POP RAX
    CMP [RSP], RAX
    MOV RAX, 0
    MOV DWORD PTR [RSP], 1
    CMOVA RAX, [RSP]
    MOV [RSP], RAX
    # End >

  POP RAX
  TEST RAX, RAX
  JZ .else_branch2

  # If then
    # DIV
    PUSH [RBP + 16] # n
    PUSH 10
    POP RAX
    XCHG [RSP], RAX
    XOR RDX, RDX
    DIV QWORD PTR [RSP]
    MOV [RSP], RAX
  CALL printHelper
  MOV [RSP], RAX

  JMP .after_else_branch2

  # If else
.else_branch2:
  PUSH 0 # Null else branch
.after_else_branch2:
  # End if
  POP RAX # Ignore non-final expression
    # ADD
    PUSH 48
      # DIV
      PUSH [RBP + 16] # n
      PUSH 10
      POP RAX
      XCHG [RSP], RAX
      XOR RDX, RDX
      DIV QWORD PTR [RSP]
      MOV [RSP], RDX
    POP RAX
    ADD [RSP], RAX
    # End ADD
  CALL printChar
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

print:
  PUSH RBP
  MOV RBP, RSP

  PUSH [RBP + 16] # n
  CALL printHelper
  MOV [RSP], RAX

  POP RAX # Ignore non-final expression
  PUSH 10
  CALL printChar
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

add:
  PUSH RBP
  MOV RBP, RSP

    # ADD
    PUSH [RBP + 24] # a
    PUSH [RBP + 16] # b
    POP RAX
    ADD [RSP], RAX
    # End ADD
  POP RAX
  POP RBP

  RET

main:
  PUSH RBP
  MOV RBP, RSP

  PUSH 1
  PUSH 2
  CALL add
  ADD RSP, 16
  PUSH RAX

  PUSH 3
  CALL add
  ADD RSP, 16
  PUSH RAX

  CALL print
  MOV [RSP], RAX

  POP RAX # Ignore non-final expression
  PUSH 1
  PUSH 2
  CALL add
  ADD RSP, 16
  PUSH RAX

  PUSH 3
  PUSH 4
  CALL add
  ADD RSP, 16
  PUSH RAX

  CALL add
  ADD RSP, 16
  PUSH RAX

  CALL print
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

_main:
  CALL main
  MOV RDI, RAX
  MOV RAX, 33554433
  SYSCALL
"
`;

exports[`compiler x86 masm print 1`] = `
"  .global _main

  .text

printChar:
  PUSH RBP
  MOV RBP, RSP

  PUSH 1
  MOV RAX, RBP
  ADD RAX, 16 # $$c
  PUSH RAX
  PUSH 1
  POP RDX
  POP RSI
  POP RDI
  MOV RAX, 33554436
  SYSCALL
  PUSH RAX

  POP RAX
  POP RBP

  RET

printHelper:
  PUSH RBP
  MOV RBP, RSP

  # If
    # >
    PUSH [RBP + 16] # n
    PUSH 9
    POP RAX
    CMP [RSP], RAX
    MOV RAX, 0
    MOV DWORD PTR [RSP], 1
    CMOVA RAX, [RSP]
    MOV [RSP], RAX
    # End >

  POP RAX
  TEST RAX, RAX
  JZ .else_branch1

  # If then
    # DIV
    PUSH [RBP + 16] # n
    PUSH 10
    POP RAX
    XCHG [RSP], RAX
    XOR RDX, RDX
    DIV QWORD PTR [RSP]
    MOV [RSP], RAX
  CALL printHelper
  MOV [RSP], RAX

  JMP .after_else_branch1

  # If else
.else_branch1:
  PUSH 0 # Null else branch
.after_else_branch1:
  # End if
  POP RAX # Ignore non-final expression
    # ADD
    PUSH 48
      # DIV
      PUSH [RBP + 16] # n
      PUSH 10
      POP RAX
      XCHG [RSP], RAX
      XOR RDX, RDX
      DIV QWORD PTR [RSP]
      MOV [RSP], RDX
    POP RAX
    ADD [RSP], RAX
    # End ADD
  CALL printChar
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

print:
  PUSH RBP
  MOV RBP, RSP

  PUSH [RBP + 16] # n
  CALL printHelper
  MOV [RSP], RAX

  POP RAX # Ignore non-final expression
  PUSH 10
  CALL printChar
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

main:
  PUSH RBP
  MOV RBP, RSP

  PUSH 231
  CALL print
  MOV [RSP], RAX

  POP RAX # Ignore non-final expression
  PUSH 4
  CALL print
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

_main:
  CALL main
  MOV RDI, RAX
  MOV RAX, 33554433
  SYSCALL
"
`;

exports[`compiler x86 masm tail-fib 1`] = `
"  .global _main

  .text

printChar:
  PUSH RBP
  MOV RBP, RSP

  PUSH 1
  MOV RAX, RBP
  ADD RAX, 16 # $$c
  PUSH RAX
  PUSH 1
  POP RDX
  POP RSI
  POP RDI
  MOV RAX, 33554436
  SYSCALL
  PUSH RAX

  POP RAX
  POP RBP

  RET

printHelper:
  PUSH RBP
  MOV RBP, RSP

  # If
    # >
    PUSH [RBP + 16] # n
    PUSH 9
    POP RAX
    CMP [RSP], RAX
    MOV RAX, 0
    MOV DWORD PTR [RSP], 1
    CMOVA RAX, [RSP]
    MOV [RSP], RAX
    # End >

  POP RAX
  TEST RAX, RAX
  JZ .else_branch5

  # If then
    # DIV
    PUSH [RBP + 16] # n
    PUSH 10
    POP RAX
    XCHG [RSP], RAX
    XOR RDX, RDX
    DIV QWORD PTR [RSP]
    MOV [RSP], RAX
  CALL printHelper
  MOV [RSP], RAX

  JMP .after_else_branch5

  # If else
.else_branch5:
  PUSH 0 # Null else branch
.after_else_branch5:
  # End if
  POP RAX # Ignore non-final expression
    # ADD
    PUSH 48
      # DIV
      PUSH [RBP + 16] # n
      PUSH 10
      POP RAX
      XCHG [RSP], RAX
      XOR RDX, RDX
      DIV QWORD PTR [RSP]
      MOV [RSP], RDX
    POP RAX
    ADD [RSP], RAX
    # End ADD
  CALL printChar
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

print:
  PUSH RBP
  MOV RBP, RSP

  PUSH [RBP + 16] # n
  CALL printHelper
  MOV [RSP], RAX

  POP RAX # Ignore non-final expression
  PUSH 10
  CALL printChar
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

tailFib:
  PUSH RBP
  MOV RBP, RSP

  # If
    # ===
    PUSH [RBP + 16] # n
    PUSH 0
    POP RAX
    CMP [RSP], RAX
    MOV RAX, 0
    MOV DWORD PTR [RSP], 1
    CMOVE RAX, [RSP]
    MOV [RSP], RAX
    # End ===

  POP RAX
  TEST RAX, RAX
  JZ .else_branch6

  # If then
  PUSH [RBP + 32] # n1
  JMP .after_else_branch6

  # If else
.else_branch6:
  PUSH [RBP + 24] # n2
    # ADD
    PUSH [RBP + 32] # n1
    PUSH [RBP + 24] # n2
    POP RAX
    ADD [RSP], RAX
    # End ADD
    # SUB
    PUSH [RBP + 16] # n
    PUSH 1
    POP RAX
    SUB [RSP], RAX
    # End SUB
  CALL tailFib
  ADD RSP, 24
  PUSH RAX

.after_else_branch6:
  # End if
  POP RAX
  POP RBP

  RET

fibHelper:
  PUSH RBP
  MOV RBP, RSP

  PUSH 0
  PUSH 1
  PUSH [RBP + 16] # n
  CALL tailFib
  ADD RSP, 24
  PUSH RAX

  POP RAX
  POP RBP

  RET

main:
  PUSH RBP
  MOV RBP, RSP

  PUSH 5
  CALL fibHelper
  MOV [RSP], RAX

  CALL print
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

_main:
  CALL main
  MOV RDI, RAX
  MOV RAX, 33554433
  SYSCALL
"
`;
